### ISSUE TITLE
Provide WebSockets Next support for authenticated JavaScript web socket upgrade and messages

### ISSUE THREAD

**Issue by sberyozkin** [quarkusio/quarkus#44409]
### Description

There are two cases that would be good to support:

1. SPA authenticates users with an OpenId Connect provider (independently of Quarkus) and sends `bearer access token` with a Web Socket upgrade request

2. Quarkus OIDC authenticates users and the authenticated users do a web socket upgrade request, with Quarkus WS Next HTTP Upgrade handler picking up an  `OIDC session cookie` - this option must fail if no CORS or CSRF protection is activated

Quarkus users have been asking how to handle these cases for years, it is really the high time, with WS Next helping users so much to do WS, offer a decent security support for native Java Script WS API calls.  

### Implementation ideas

For the SPA sending a bearer token, using `Secure-WebSocket-Protocol` is a defacto work-around, we should offer a support for this option, I've found https://ably.com/blog/websocket-authentication informative, and this [Kubernetes PR](https://github.com/kubernetes/kubernetes/commit/714f97d7baf4975ad3aa47735a868a81a984d1f0) specifically. 

For the user with the OIDC session cookie, we have to enforce CORS. I'd like to demo this case with the Quarkus LangChain4j secure chat bot demo.

I suppose in both cases we should have a custom WS HTTP Upgrade handler which will use OIDC TenantIdentityProvider to verify the identity... 

I'm not sure how to structure it correctly, I can think of 2 options:

1. WS-Next introduces an SPI which `quarkus-oidc` will depend upon and will  implement a custom HTTP Upgrade handler.
2. We have a very light `quarkus-oidc-websockets-next` extension which only ships a custom HTTP Upgrade handler

CC to the WS experts: @mkouba @michalvavrik

**Comment by sberyozkin** [comment]
@michalvavrik @mkouba The cookie based system already works since the session cookie is verified.

**Comment by sberyozkin** [comment]
@mkouba @michalvavrik I continue playing with the second option so far and it works well. 

Does WebSockets client (any type of client, for example, the one created with JavaScript) pass some kind of  key/id with every message it sends to the server, where this key/id was provided to it by the server when the upgrade was completed ?

Specifically, when the client sends a new message, I'd like to assert that this key/id matches a corresponding `SecurityIdentity`'s attribute.

**Comment by sberyozkin** [comment]
@mkouba @michalvavrik  What would be great to have is for `HttpUpgradeContext` provide some unique value that is available at the upgrade time and which is going to be also provided with the subsequent client messages, internally... Then we could record this unique context property as a security identity attribute to be 100% sure that subsequent WSS client messages are bound to this security identity... 

Right now, we know the security identity is bound to the connection but we don't have extra assurances...

The other thing which would be nice to have is to be able, possibly via a WS extension, to return the token or encrypted session cookie back to the WS client, so that the `re-authentication` is done on every WS client message

**Comment by michalvavrik** [comment]
Sorry @sberyozkin , ATM I am bit overwhelmed with openend PRs & work & feeling sick so I can't read / answer. I'll read it on Friday if it will be still relevant for you. Thanks

**Comment by sberyozkin** [comment]
@michalvavrik Np at all, on the upgrade path we actually have everything with `HttpUpgradeCheck` to make it really secure, for ex, with a custom ticket system, which I'll do in the demo...

But having an option to retain an indirect security identity key on the client side even after the upgrade would be great, not a blocker, but would be nice to have

**Comment by sberyozkin** [comment]
The reason I'd like to have a WS client per-request re-authentication eventually, and for it to work, we should be able to retain an original token or cookie with the WS client, is to support, among other things, in case of the authorization code flow, an access token propagation....

**Comment by mkouba** [comment]
> Does WebSockets client (any type of client, for example, the one created with JavaScript) pass some kind of key/id with every message it sends to the server, where this key/id was provided to it by the server when the upgrade was completed ?
>

Not really. In WS no metadata are sent with a message AFAIK. So the only option is to include this unique identifier in the message payload. But then you would need to understand the payload which is not the case for WS Next.

> What would be great to have is for HttpUpgradeContext provide some unique value that is available at the upgrade time and which is going to be also provided with the subsequent client messages, internally...

You can generate some unique id and put the value in a response header - see `HttpUpgradeCheck.CheckResult`. But even then we're where we were because there are no metadata in WS messages.

**Comment by sberyozkin** [comment]
Thanks @mkouba, I'm happy enough that `HttpUpgradeCheck` can accommodate any security schemes in making sure the upgrade request came in from the authenticated user, for ex, with a custom ticket.

For this specific issue, handling a `Secure-WebSocket-Protocol` workaround to extract a bearer token to make it available for the verification during the upgrade is what we can do...
 
As far as forwarding extra pieces of metadata with the message payload is concerned... I thought this is what WS extensions can handle, for example, if it is a text message, and say we have a `text-with-metadata` extension, then when the Java script WS API sends text like `hello` what goes into the channel is JSON, with a `metadata` property and `text` property, and then on the server side, the text is available to the WS endpoint, and the metadata - to registered message metadata handlers.... 

Does it sound too unrealistic or we can consider something like that ?

**Comment by sberyozkin** [comment]
I.e, you are absolutely right that it goes with the payload, but the registered extension 1) unwraps the text 2) lets metadata handlers handle metadata

**Comment by michalvavrik** [comment]
1. I had a chance to re-read whole conversation and the part I am missing is how do you get credentials from [JS WS client](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_client_applications) without sending it in the payload? You do not control the client side, so you can never alter the payload unless you agree with users on what payload structure will look like.

2. https://www.rfc-editor.org/rfc/rfc6455.html says:
 _The WebSocket
   server can use any client authentication mechanism available to a
   generic HTTP server, such as cookies, HTTP authentication, or TLS
   authentication._
  I found other WS libraries for WebSockets, intentionally not listing them here because I don't know which one I can trust in the 
  production, but some offers support for headers. It seems to me like WS security is origin-based and we should prioritize WSS protocol support. Or maybe that already works? Sorry, I didn't check :weary:

> The reason I'd like to have a WS client per-request re-authentication eventually, and for it to work, we should be able to retain an original token or cookie with the WS client, is to support, among other things, in case of the authorization code flow, an access token propagation....

3. I don't see an option to use cookies with WS JS client here https://developer.mozilla.org/en-US/docs/Web/API/WebSocket, how would you propagate CSRF-Tokens without cookies or headers? WebSockets are not secured with the same-origin policy.

4. I agree it would be fantastic to provide users Quarkus-way to use security with WS Next and JS WS client, but I think you are heading either to custom [extension](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/extensions) or custom [protocol](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/protocol). It will be a lot of workaround without knowing if users cannot use different clients that support headers for the initial handshake?
